libyul/optimiser/OptimizerUtilities.cpp

```diff
@@ -37,3 +42,15 @@ void yul::removeEmptyBlocks(Block& _block)
 	};
 	boost::range::remove_erase_if(_block.statements, isEmptyBlock);
 }
+
+// TODO: Should this be part of the Dialect instead?
+bool yul::isRestrictedIdentifier(Dialect const& _dialect, YulString const& _identifier)
+{
+	if (TokenTraits::keywordByName(_identifier.str()) != Token::Identifier)
+		return true;
```

Comment:
This now *does* relate to what I wrote earlier - this will suffer the same issue of it being the solidity tokens as ``AsmParser.cpp``, doesn't it?
So - ugly as this may be - this needs to exclude things like ``Token::Var`` and ``Token::Address``.
We should probably add a ``TokenTraits::isYulToken``...

Reply:
If we do not move it to be a helper, then we could leave it as it is, couldn't we?

Reply:
I'm not sure I understand... what I mean is that e.g. this test case will break:
```
{
  let address_1 := 1
}
// ----
// step: varNameCleaner
//
// {
//     let address := 1
// }
```

Reply:
Why would it break? `address` is marked as restricted, e.g it won't rename it to it.

Reply:
But it *should* :-)!

Reply:
``address`` is a perfectly fine identifier in yul (see the workarounds in ``AsmParser.cpp``). So it should *not* be considered restricted.

Reply:
`address` is also an EVM opcode, so incidentally should not be possible, at least in EVMDialect.

Reply:
Ok, bad example - then ``var`` and ``in`` :-).

Reply:
And ``address`` only in the sense that it should rather be caught a few lines down.
But e.g. ``in`` and ``var`` are both actually produced by IR codegen (that's why I know this issue in the first place).

Reply:
This can be simplified/finished once #9331 is merged.